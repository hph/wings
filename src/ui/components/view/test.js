import React from 'react';
import renderer from 'react-test-renderer';
import _ from 'lodash';

import { View } from './index';

jest.mock('../cursor', () => 'Cursor');
jest.mock('../line-numbers', () => 'LineNumbers');

function createNodeMock (element) {
  if (element.type === 'div') {
    return {
      offsetWidth: 0,
      offsetHeight: 0,
    };
  }

  return null;
}

const defaultProps = {
  config: {
    charHeight: 20,
    charWidth: 5,
    showLineNumbers: false,
  },
  splits: 1,
  views: [],
  dispatch: () => {},
  isFirst: true,
  view: {
    id: 1,
    height: 800,
    lines: [],
    firstVisibleRow: 0,
    firstVisibleColumn: 0,
  },
  viewId: 1,
};

const createSnapshot = (props = defaultProps) => {
  expect(
    renderer.create(<View {...props} />, { createNodeMock },
  ).toJSON()).toMatchSnapshot();
};

test('View renders empty content', () => {
  createSnapshot();
});

test('View renders a Cursor in the current view', () => {
  createSnapshot({
    ...defaultProps,
    config: {
      ...defaultProps.config,
      currentViewId: 1,
    },
  });
});

test('View renders LineNumbers when so configured', () => {
  createSnapshot({
    ...defaultProps,
    config: {
      ...defaultProps.config,
      showLineNumbers: true,
    },
  });
});

test('View renders lines of code', () => {
  createSnapshot({
    ...defaultProps,
    view: {
      ...defaultProps.view,
      lines: [
        'hello, world!',
        'this is just marvellous!',
      ],
    },
  });
});

test('View only renders visible lines', () => {
  // Since the line height is set to 20, and the configured view height is 100,
  // we will render 5 (100/20) lines, not the full 100.
  createSnapshot({
    ...defaultProps,
    view: {
      ...defaultProps.view,
      lines: _.range(100),
      height: 100,
    },
  });
});

test('View starts rendering lines from the first visible row as configured', () => {
  createSnapshot({
    ...defaultProps,
    view: {
      ...defaultProps.view,
      lines: _.range(100),
      height: 100,
      firstVisibleRow: 10,
    },
  });
});

test('View width is determined by the number of splits', () => {
  const createComponent = splits => {
    return renderer.create(
      <View {...defaultProps} splits={splits} />, { createNodeMock },
    ).toJSON();
  };
  expect(createComponent(1).props.style).toEqual({ width: '100%' });
  expect(createComponent(2).props.style).toEqual({ width: '50%' });
  expect(createComponent(3).props.style).toEqual({ width: `${ 100 / 3 }%` });
});

test('View renders a border when it is not the first and there are no line numbers', () => {
  createSnapshot({
    ...defaultProps,
    isFirst: false,
    config: {
      ...defaultProps.config,
      showLineNumbers: false,
    },
  });
  createSnapshot({
    ...defaultProps,
    isFirst: false,
    config: {
      ...defaultProps.config,
      showLineNumbers: true,
    },
  });
});

test('View has a left offset as determined by the first visible column', () => {
  createSnapshot({
    ...defaultProps,
    view: {
      ...defaultProps.view,
      firstVisibleColumn: 5,
    },
  });
});
