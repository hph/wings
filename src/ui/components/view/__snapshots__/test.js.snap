// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View has a left offset as determined by the first visible column 1`] = `
<div
  className="root overlay"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <div
    className="text"
    onMouseDown={[Function]}
    style={
      Object {
        "left": -25,
      }
    }
  />
</div>
`;

exports[`View only renders visible lines 1`] = `
<div
  className="root overlay"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <div
    className="text"
    onMouseDown={[Function]}
    style={
      Object {
        "left": -0,
      }
    }
  >
    <div>
      0
    </div>
    <div>
      1
    </div>
    <div>
      2
    </div>
    <div>
      3
    </div>
    <div>
      4
    </div>
  </div>
</div>
`;

exports[`View renders LineNumbers when so configured 1`] = `
<div
  className="root overlay"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <LineNumbers
    className="numbers"
    config={
      Object {
        "charHeight": 20,
        "charWidth": 5,
        "showLineNumbers": true,
      }
    }
    innerRef={[Function]}
    view={
      Object {
        "firstVisibleColumn": 0,
        "firstVisibleRow": 0,
        "height": 800,
        "id": 1,
        "lines": Array [],
      }
    }
  />
  <div
    className="text"
    onMouseDown={[Function]}
    style={
      Object {
        "left": -0,
      }
    }
  />
</div>
`;

exports[`View renders a Cursor in the current view 1`] = `
<div
  className="root overlay"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <div
    className="text"
    onMouseDown={[Function]}
    style={
      Object {
        "left": -0,
      }
    }
  >
    <Cursor
      config={
        Object {
          "charHeight": 20,
          "charWidth": 5,
          "currentViewId": 1,
          "showLineNumbers": false,
        }
      }
      view={
        Object {
          "firstVisibleColumn": 0,
          "firstVisibleRow": 0,
          "height": 800,
          "id": 1,
          "lines": Array [],
        }
      }
    />
  </div>
</div>
`;

exports[`View renders a border when it is not the first and there are no line numbers 1`] = `
<div
  className="root overlay"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <div
    className="text border"
    onMouseDown={[Function]}
    style={
      Object {
        "left": -0,
      }
    }
  />
</div>
`;

exports[`View renders a border when it is not the first and there are no line numbers 2`] = `
<div
  className="root overlay"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <LineNumbers
    className="numbers"
    config={
      Object {
        "charHeight": 20,
        "charWidth": 5,
        "showLineNumbers": true,
      }
    }
    innerRef={[Function]}
    view={
      Object {
        "firstVisibleColumn": 0,
        "firstVisibleRow": 0,
        "height": 800,
        "id": 1,
        "lines": Array [],
      }
    }
  />
  <div
    className="text"
    onMouseDown={[Function]}
    style={
      Object {
        "left": -0,
      }
    }
  />
</div>
`;

exports[`View renders empty content 1`] = `
<div
  className="root overlay"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <div
    className="text"
    onMouseDown={[Function]}
    style={
      Object {
        "left": -0,
      }
    }
  />
</div>
`;

exports[`View renders lines of code 1`] = `
<div
  className="root overlay"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <div
    className="text"
    onMouseDown={[Function]}
    style={
      Object {
        "left": -0,
      }
    }
  >
    <div>
      hello, world!
    </div>
    <div>
      this is just marvellous!
    </div>
  </div>
</div>
`;

exports[`View starts rendering lines from the first visible row as configured 1`] = `
<div
  className="root overlay"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <div
    className="text"
    onMouseDown={[Function]}
    style={
      Object {
        "left": -0,
      }
    }
  >
    <div>
      10
    </div>
    <div>
      11
    </div>
    <div>
      12
    </div>
    <div>
      13
    </div>
    <div>
      14
    </div>
  </div>
</div>
`;
